src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py

```diff
@@ -0,0 +1,234 @@
+from pathlib import Path
+from typing import Callable, List, Optional
+
+import numpy as np
+import torch
+
+from ffcv.fields.basics import IntDecoder
+from ffcv.loader import OrderOption
+from sparseml.pytorch.utils import default_device
+
+
+try:
+    import ffcv
+    from ffcv import DatasetWriter, Loader
+    from ffcv.fields import IntField, RGBImageField
+    from ffcv.fields.rgb_image import (
+        CenterCropRGBImageDecoder,
+        RandomResizedCropRGBImageDecoder,
+    )
+    from ffcv.pipeline.operation import Operation
+    from ffcv.transforms import (
+        NormalizeImage,
+        RandomHorizontalFlip,
+        Squeeze,
+        ToDevice,
+        ToTensor,
+        ToTorchImage,
+    )
+except Exception as ffcv_error:
+    ffcv = None
+    ffcv_error = ffcv_error
+
+LOGGER = logging.getLogger(__name__)
+
+
+class FFCVDataset(ABC):
+    """
+    Abstract class for FFCV datasets.
+    """
+
+    def __init__(self, dataset, write_path: str = "data/ffcv/", *args, **kwargs):
+        """
+        Initialize the dataset.
+        """
+        self.dataset = dataset
+        self.write_path = write_path
+        Path(self.write_path).parent.mkdir(parents=True, exist_ok=True)
+
+    @abstractmethod
+    def _write(self, *args, **kwargs):
+        """
+        Write the dataset to disk.
+        """
+        pass
+
+    @abstractmethod
+    def switched_off_transform(self, *args, **kwargs):
+        """
+        Switch off the transform.
+        """
+        pass
+
+
+class ImageFolderFFCV(FFCVDataset):
+    """
+    Helper class using FFCV for ImageFolder type datasets.
+    """
+
+    IMAGENET_MEAN = np.array([0.485, 0.456, 0.406]) * 255
+    IMAGENET_STD = np.array([0.229, 0.224, 0.225]) * 255
+    DEFAULT_CROP_RATIO = 224 / 256
+
+    def __init__(
+        self,
+        dataset,
+        write_path: str = "data/ffcv.beton",
+        transform: Optional[Callable] = None,
+        num_workers: int = 16,
+        image_size: int = 224,
+        device=default_device(),
+    ):
+        """
+        Initialize the dataset.
+        """
+        super().__init__(
+            dataset=dataset,
+            write_path=write_path,
+        )
+        self.transform = transform
+
+        with self.switched_off_transform():
+            self._write(num_workers=num_workers)
+
+        self._image_size = image_size
+        self._device = device
+
+    def _write(self, num_workers: int = 16):
+        """
+        Writes the dataset to the write_path.
+        """
+        assert ffcv, "FFCV is not installed"
+        writer = DatasetWriter(
+            self.write_path,
+            {
+                "image": RGBImageField(),
+                "label": IntField(),
+            },
+            num_workers=num_workers,
+        )
+
+        LOGGER.info(f"Writing dataset to {self.write_path}")
+        writer.from_indexed_dataset(self.dataset)
+
+    @contextmanager
+    def switched_off_transform(self):
```

Comment:
let's move all this into write in the parent class since write will be a concrete method